tfx_bsl/beam/bsl_util.py

```diff
@@ -0,0 +1,132 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""TFX-BSL util"""
+
+from __future__ import absolute_import
+from __future__ import division
+# Standard __future__ imports
+from __future__ import print_function
+
+import numpy as np
+import pyarrow as pa
+import pandas as pd
+import base64
+import json
+import typing
+from typing import Dict, List, Text, Any, Set, Mapping, Optional
+from tfx_bsl.beam.bsl_constants import _RECORDBATCH_COLUMN
+
+_KERAS_INPUT_SUFFIX = '_input'
+
+def ExtractSerializedExampleFromRecordBatch(elements: pa.RecordBatch) -> List[Text]:
+  serialized_examples = None
+  for column_name, column_array in zip(elements.schema.names, elements.columns):
+    if column_name == _RECORDBATCH_COLUMN:
+      column_type = column_array.flatten().type
+      if not (pa.types.is_binary(column_type) or pa.types.is_string(column_type)):
+        raise ValueError(
+          'Expected a list of serialized examples in bytes or as a string, got %s' %
+          type(example))
+      serialized_examples = column_array.flatten().to_pylist()
+      break
+
+  if not serialized_examples:
+    raise ValueError('Raw examples not found.')
+
+  return serialized_examples
+
+
+def RecordToJSON(
+  record_batch: pa.RecordBatch, prepare_instances_serialized) -> List[Mapping[Text, Any]]:
+  """Returns a list of JSON dictionaries translated from `record_batch`.
+
+    The conversion will take in a recordbatch that contains features from a 
+    tf.train.Example and will return a list of dict like string (JSON) where 
+    each item is a JSON representation of an example.
+
+    Return:
+      List of JSON dictionaries
+      - format: [{ feature1: value1, feature2: [value2_1, value2_2]... }, ...]
+
+  Args:
+  record_batch: input RecordBatch.
+  """
+
+  # TODO (b/155912552): Handle this for sequence example.
+
+  def flatten(element: List[Any]):
+    if len(element) == 1:
+        return element[0]
+    return element
+
+  df = record_batch.to_pandas()
+  if prepare_instances_serialized: 
+    return [{'b64': base64.b64encode(value).decode()} for value in df[_RECORDBATCH_COLUMN]]
+  else:
+    as_binary = df.columns.str.endswith("_bytes")
```

Comment:
Why does the name end with "_bytes"?

Reply:
User specified byte columns, it's consistent with the original implementation

Reply:
This is required by cloud ai platform to indicate the bytes feature with '_bytes' suffix.